h5py/h5fd.pyx

```diff
@@ -77,3 +77,79 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
 # Flag for tracking allocation of space in file
 LOG_ALLOC     = H5FD_LOG_ALLOC      # 0x4000
 LOG_ALL       = H5FD_LOG_ALL        # (H5FD_LOG_ALLOC|H5FD_LOG_TIME_IO|H5FD_LOG_NUM_IO|H5FD_LOG_FLAVOR|H5FD_LOG_FILE_IO|H5FD_LOG_LOC_IO)
+
+
+# H5FD_t of file-like object
+ctypedef struct H5FD_fileobj_t:
+  H5FD_t base
+  PyObject* fileobj
+  haddr_t eoa
+
+
+from cpython cimport Py_INCREF, Py_DECREF
+from libc.stdint cimport *
+from libc.stdio cimport *
+
+
+cdef H5FD_fileobj_t *H5FD_fileobj_open(const char *name, unsigned flags, hid_t fapl, haddr_t maxaddr):
+    cdef H5FD_fileobj_t *f = <H5FD_fileobj_t *>malloc(sizeof(H5FD_fileobj_t))
+    cdef uintptr_t p
+    sscanf(name, "%lx", &p)
+    f.fileobj = <PyObject*>p
+    Py_INCREF(<object>f.fileobj)
+    f.eoa = 0
+    return f
+
+cdef herr_t H5FD_fileobj_close(H5FD_fileobj_t *f):
+    Py_DECREF(<object>f.fileobj)
+    return 0
+
+cdef haddr_t H5FD_fileobj_get_eoa(const H5FD_fileobj_t *f, H5FD_mem_t type):
+    return f.eoa
+
+cdef herr_t H5FD_fileobj_set_eoa(H5FD_fileobj_t *f, H5FD_mem_t type, haddr_t addr):
+    f.eoa = addr
+    return 0
+
+cdef haddr_t H5FD_fileobj_get_eof(const H5FD_fileobj_t *f, H5FD_mem_t type):
+    (<object>f.fileobj).seek(0, SEEK_END)
+    return (<object>f.fileobj).tell()
+
+cdef herr_t H5FD_fileobj_read(H5FD_fileobj_t *f, H5FD_mem_t type, hid_t dxpl, haddr_t addr, size_t size, void *buf):
+    (<object>f.fileobj).seek(addr)
+    cdef b = (<object>f.fileobj).read(size)
+    cdef const unsigned char[:] mview = b
+    memcpy(buf, &mview[0], size)
+    return 0
+
+cdef herr_t H5FD_fileobj_write(H5FD_fileobj_t *f, H5FD_mem_t type, hid_t dxpl, haddr_t addr, size_t size, const void *buf):
+    (<object>f.fileobj).seek(addr)
+    (<object>f.fileobj).write(bytes(<const unsigned char[:size]>buf))
+    return 0
+
+cdef herr_t H5FD_fileobj_truncate(H5FD_fileobj_t *f, hid_t dxpl, hbool_t closing):
+    (<object>f.fileobj).truncate(f.eoa)
+    return 0
+
+cdef herr_t H5FD_fileobj_flush(H5FD_fileobj_t *f, hid_t dxpl, hbool_t closing):
+    (<object>f.fileobj).flush()
+    return 0
+
+
+cdef H5FD_class_t info
+memset(&info, 0, sizeof(info))
+
+info.name = 'fileobj'
```

Comment:
I'm confused what's going on in this section, can we add some docs outlining what's going on here?

Reply:
Sure, I'll do. In short, I'm allocating a VFL driver descriptor that dispatches HDF5 I/O to `H5FD_fileobj*` procedures (which, in turn, translate that to Python file-like object method calls).